label_studio/io_storages/aperturedb/models.py

```diff
@@ -0,0 +1,557 @@
+        self._attr_must_match(self.rect, other.rect, "y")
+        self._attr_must_match(self.rect, other.rect, "width")
+        self._attr_must_match(self.rect, other.rect, "height")
+        self.labels.extend(other.labels)
+        if len(self.labels) > 1:
+            raise ValueError(
+                f"rectangle annotation can have at most one label: {','.join(self.labels)}")
+        self.text.extend(other.text)
+
+
+class ApertureDBExportStorage(ApertureDBStorageMixin, ExportStorage):
+    def save_annotation(self, annotation):
+        db = self.get_connection()
+        logger.debug(
+            f"Creating new object on {self.__class__.__name__} Storage {self} for annotation {annotation}...")
+        ser_annotation = self._get_serialized_data(annotation)
+
+        bbox_map = {}
+        regions = annotation.result if isinstance(
+            annotation.result, list) else [annotation.result]
+        for ann in regions:
+            if not "type" in ann:
+                continue
+
+            if ann["id"] in bbox_map:
+                bbox_map[ann["id"]].merge(ann)
+            elif ann["type"].startswith("rectangle"):
+                bbox_map[ann["id"]] = AnnotationBBox(ann)
+
+        img_id = annotation.task.get_storage_filename()
+        if not img_id:
+            img_url = annotation.task.data["image"]
+            img_id = re.split("key=", img_url)[-1]
+
+        ann_id = str(annotation.id)
+
+        # Initial scan to see if bounding boxes already exist.
+        # This is necessary because some existing bounding boxes don't have LS_id set.
+        # Instead we rely on the LS id being the same as the _uniqueid.
+        query = [
+            {
+                "FindImage": {
+                    "blobs": False,
+                    "constraints": {"_uniqueid": ["==", img_id]},
+                    "_ref": 1,
+                }
+            },
+            {
+                "FindBoundingBox": {
+                    "blobs": False,
+                    "image_ref": 1,
+                    "results": {"list": ["LS_id", "_uniqueid"]},
+                }
+            }
+        ]
+        res, _ = db.query(query)
+        status = self._response_status(res)
+        if status != 0:
+            raise ValueError(
+                f"Error saving annotation data to ApertureDB : {db.get_last_response_str()}")
+        # Scan results to see if they exist or not, and which field to use as the id.
+        bbox_id_fields = {}
+        for bbx in res[1]["FindBoundingBox"]["entities"]:
+            if bbx["LS_id"] in bbox_map:
+                bbox_id_fields[bbx["LS_id"]] = "LS_id"
+            elif bbx["_uniqueid"] in bbox_map:
+                bbox_id_fields[bbx["_uniqueid"]] = "_uniqueid"
+
+        logger.debug(f"{img_id=}, {ann_id=}, {bbox_id_fields=}")
+
+        # TODO: Small race condition here. If a bounding box is added between the initial scan and the save, it will be missed.
```

Comment:
in theory we could get the unique ids of all the bounding boxes in the initial scan and verify afterwards that they match, no?

Reply:
Yes, but I think there are other possible race conditions (that I did not list) that could not be detected that way.